### PR TITLE
[RLlib] RLModule API change: If "actions" key returned from forward_inference|exploration, use actions as-is.

### PR DESCRIPTION
**Pull request by sven1977**
RLModule API change:
* If "actions" key returned from forward_inference|exploration, use actions as-is.
* If - additionally - `action_dist_inputs` are returned, use these to construct a distribution and compute action probs/logs from it.
* If "actions" key is NOT returned, continue using "action_dist_inputs" (no change wrt current behavior).

Changed the docs to reflect this.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/rllib/rllib-rlmodule.rst [hunk 1]
```diff
@@ -426,7 +426,26 @@ What your customization could have looked like before:
                     return None, None, None
 
 
-All of the ``Policy.compute_***`` functions expect that `~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration` and `~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference` return a dictionary that contains the key "action_dist_inputs", whose value are the parameters (inputs) of a ``ray.rllib.models.distributions.Distribution`` class. Commonly used distribution implementations can be found under ``ray.rllib.models.tf.tf_distributions`` for tensorflow and ``ray.rllib.models.torch.torch_distributions`` for torch. You can choose to return determinstic actions, by creating a determinstic distribution instance. See `Writing Custom Single Agent RL Modules`_ for more details on how to implement your own custom RL Module.
+All of the ``Policy.compute_***`` functions expect that
+`~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration` and `~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference`
+return a dictionary that either contains the key "actions" and/or the key "action_dist_inputs".
+
+If you return the "actions" key:
+* RLlib will use the actions provided thereunder directly and as-is.
```
Comment (kouroshHakha) [inline doc/source/rllib/rllib-rlmodule.rst:434]: The formatting is broken (these are not rendered as bullet points)

https://anyscale-ray--36067.com.readthedocs.build/en/36067/rllib/rllib-rlmodule.html#migrating-from-custom-policies-and-models-to-rl-modules

File: doc/source/rllib/rllib-rlmodule.rst [hunk 2]
```diff
@@ -426,7 +426,26 @@ What your customization could have looked like before:
...
+* If you also returned the "action_dist_inputs" key: RLlib will also create a ``ray.rllib.models.distributions.Distribution`` object from the distribution parameters under that key and - in the case of ``forward_exploration()`` - compute action probs and logp values from the given actions automatically.
+
+If you do not return the "actions" key:
+* You must return the "action_dist_inputs" key instead from your ``forward_inference()`` and ``forward_exploration()`` methods.
+* RLlib will create a ``ray.rllib.models.distributions.Distribution`` object from the distribution parameters under that key and sample actions from the thus generated distribution.
+* In the case of ``forward_exploration()``, RLlib will also compute action probs and logp values from the sampled actions automatically.
+
+Note that in the case of ``forward_inference()``, the generated distributions (from returned key "action_dist_inputs") will always be made deterministic via
```
Comment (kouroshHakha) [inline doc/source/rllib/rllib-rlmodule.rst:442]: Let's make this a Note box via `.. note::`?

File: doc/source/rllib/rllib-rlmodule.rst [hunk 3]
```diff
@@ -426,7 +426,26 @@ What your customization could have looked like before:
...
+`~ray.rllib.core.rl_module.rl_module.RLModule.forward_exploration` and `~ray.rllib.core.rl_module.rl_module.RLModule.forward_inference`
+return a dictionary that either contains the key "actions" and/or the key "action_dist_inputs".
+
+If you return the "actions" key:
+* RLlib will use the actions provided thereunder directly and as-is.
+* If you also returned the "action_dist_inputs" key: RLlib will also create a ``ray.rllib.models.distributions.Distribution`` object from the distribution parameters under that key and - in the case of ``forward_exploration()`` - compute action probs and logp values from the given actions automatically.
+
+If you do not return the "actions" key:
+* You must return the "action_dist_inputs" key instead from your ``forward_inference()`` and ``forward_exploration()`` methods.
+* RLlib will create a ``ray.rllib.models.distributions.Distribution`` object from the distribution parameters under that key and sample actions from the thus generated distribution.
```
Comment (kouroshHakha) [inline doc/source/rllib/rllib-rlmodule.rst:439]: All of these classes we mention here have to be linked to their class definition API reference. You can use something like `:py:class:~ray.rllib.models.distributions.Distribution` to achieve this.

File: doc/source/rllib/rllib-rlmodule.rst
Comment (kouroshHakha) [inline doc/source/rllib/rllib-rlmodule.rst:1]: This added documentation is not specific to those who are migrating from policy API. We should put it under the right section. I think adding it somewhere close to [Writing Custom Single Agent RL Modules](https://anyscale-ray--36067.com.readthedocs.build/en/36067/rllib/rllib-rlmodule.html#writing-custom-single-agent-rl-modules) would be the way to go (it requires getting rid of those policy specific numenclature)

Maybe we can consolidate your paragraph with the description of what needs to be implemented for each forward method?  

image

**Review by kouroshHakha — COMMENTED** [review]
Implementation looks good. We need to clean up the docs a little bit. Left some comments on my concerns. 

Also I think these details should be now reflected as part of the RLModule API references under the three forward methods. That's probably where the advanced users will look for information on implementation details the most. 
We need to be more clear on what the output will be treated in the doc-string of those functions (under Returns section)

**Comment by sven1977** [comment]
Thanks for your review @kouroshHakha ! All very valid concerns and suggestions. I'll make these changes :)

File: doc/source/rllib/rllib-rlmodule.rst (hunk 2 above)
Comment (sven1977) [reply doc/source/rllib/rllib-rlmodule.rst:442]: done

File: doc/source/rllib/rllib-rlmodule.rst (hunk 3 above)
Comment (sven1977) [reply doc/source/rllib/rllib-rlmodule.rst:439]: done

File: doc/source/rllib/rllib-rlmodule.rst (hunk 1 above)
Comment (sven1977) [reply doc/source/rllib/rllib-rlmodule.rst:434]: Fixed.

File: doc/source/rllib/rllib-rlmodule.rst (hunk 3 above)
Comment (sven1977) [reply doc/source/rllib/rllib-rlmodule.rst:439]: fixed all of these via `:py:meth:..` and `:py:class:..` links.

**Review by kouroshHakha — APPROVED** [review]

**Pull request merged**